packages/amplify-cli-core/src/state-manager/stateManager.ts

```diff
@@ -72,9 +75,129 @@ export class StateManager {
+    const { envName } = this.getLocalEnvInfo()
+
+    //validate if the policies match the custom IAM policies schema, if not, then not write into the CFN template
+    const validatePolicies = ajv.compile(CustomIAMPoliciesSchema);
+    if (!data || !validatePolicies(data)) {
+      printer.warn(`The format of custom IAM policies in the ${categoryName} ${resourceName} is not valid`);
+      return undefined;
+    }
+    const validatePolicy = ajv.compile(CustomIAMPolicySchema);
+    for (const policy of data.policies) {
+      if (validatePolicy(policy)) {
+        if (!policy.Effect) policy.Effect = 'Allow';
+        const policyWithoutEnv = this.replaceEnvForCustomPolicies(policy, envName);
+        customPolicies.policies.push(policyWithoutEnv);
+      }
+      else {
+        printer.warn(`The format of custom IAM policies in the ${categoryName} ${resourceName} is not valid`);
+        return undefined;
+      }
+    }
+    //the policies with env will be written to the custom policies file again
+    JSONUtilities.writeJson(filePath, data);
+    //the policies without env will be carried over and merge into the CFN template
+    return customPolicies;
+  };
+
+  //replace or add env parameter in the front of the resource customers enter to the current env
+  replaceEnvForCustomPolicies = (policy: CustomIAMPolicy, currentEnv: string): CustomIAMPolicy => {
+    let action = policy.Action;
+    let effect = policy.Effect;
+    let resourceWithNoEnv = [];
+    let resourceWithEnv = [];
+
+    for (let resource of policy.Resource){
+      const splitedResource = resource.split('/');
+      //if there is no env in the resource customers entered, then just add the current env to the resource
+      if(splitedResource.length <= 1) {
+        resourceWithEnv.push(`${currentEnv}/${splitedResource[0]}`);
+        resourceWithNoEnv.push(splitedResource[0]);
+        continue;
+      }
+      //In case there is a slash in the env name, check where is the start of the ARN
+      let index = 0;
+      for (index; index < splitedResource.length; index ++) {
+        if (splitedResource[index].substring(0,3) === 'arn') {
+          break;
+        }
+      }
+      //change the env if env exists, or add the env if not exists.
+      if (index >= 1) {
+        let env = '';
+        for (let i = 0; i < index; i++) {
+          env = `${env}${splitedResource[i]}`;
+          if (i < (index - 1)) env = `${env}/`
+        }
+        if (env != currentEnv) env = currentEnv;
+        let combinedResourceWithEnv = env;
+        for (let j = index; j < splitedResource.length; j++) {
+          combinedResourceWithEnv = `${combinedResourceWithEnv}/${splitedResource[j]}`
+        }
+        resourceWithEnv.push(combinedResourceWithEnv);
+        splitedResource[0] = '';
+        resourceWithNoEnv.push(splitedResource.join(''));
+      } else {
+        resourceWithEnv.push(`${currentEnv}/${resource}`);
+        resourceWithNoEnv.push(resource);
+      }
+    }
```

Comment:
`"arn::{env}:abc".replace({env}, envName)` you can use a string replace function, I am not exactly sure what this is doing

Reply:
In the design doc, the resource is "<env>/arn:aws:s3:::my_bucket/*", I am trying to get the env from the string, and replace it if it's not the same as the current env. And in case customers put the slash in the resource, I do some verification on it. The resource without env will be used to update CFN template, and the resource with env will be written into the custom policies file again

Reply:
I think the design doc says the customer will provide `{env}` variable in the resource arn and you just have to replace it in code

Reply:
But if customer is in the "dev" environment now, and they provide "prod" in the resource. If we don't correct it in the custom-policies.json, they will still see the "prod" in the resource, however it's being use in "dev". Wouldn't it be a bit confused?

Reply:
You won't need to write it back you are just supposed to replace it in memory without comitting it to a file and then write it to the CFN